File: app/assets/javascripts/recipes/models/recipe.js
import _ from 'lodash';

export default class Recipe {
  constructor(name, ingredientsList, preparationStepsList, kcal=0, portionsQuantity=0, sourceUrl=undefined) {
    this.name = name;
    this.ingredientsList = ingredientsList;
    this.preparationStepsList = preparationStepsList;
    this.kcal = kcal;
    this.portionsQuantity = portionsQuantity;
    this.sourceUrl = sourceUrl
  }

  toParams = () => {
    return({
      name: this.name,
      ingredients: {main: this.ingredientsListParams()},
      preparation_steps: this.preparationStepsListParams()
    })
  };

  ingredientsListParams = () => {
    return _
      .chain(this.ingredientsList)
      .map((ingredient, ingredientOrder) => {
        return {
          order_number: this.takeNumber(ingredientOrder),
          name: ingredient.name
        }
      })
      .values()
      .value();
  };

  preparationStepsListParams = () => {
    return _
      .chain(this.preparationStepsList)
      .map((preparationStep, preparationStepOrder) => {
        return {
          order_number: this.takeNumber(preparationStepOrder),
          description: preparationStep.description
        }
      })
      .values()
      .value()
  };

  takeNumber = (string) => {
    return string.match(/\d+/)[0]
  };
};

File: app/models/recipe.rb
class Recipe < ActiveRecord::Base

  def ingredients=(ingredient_sections)
    list = ingredient_sections.each_with_object({}) do |(name, ingredients), ingredient_list|
      ingredient_list[name] = ingredients.map(&:to_hash)
    end
    write_attribute(:ingredients, YAML.dump(list))
  end

  def ingredients
    YAML.load(read_attribute(:ingredients)).each_with_object({}) do |(name, ingredients), ingredients_list|
      ingredients_list[name] = ingredients.map do |ingredient|
        Ingredient.new(ingredient[:name])
      end
    end
  end

  def preparation_steps=(preparation_steps)
    hashed_preparation_steps = YAML.dump(preparation_steps.map(&:to_hash))
    write_attribute(:preparation_steps, hashed_preparation_steps)
  end

  def preparation_steps
    YAML.load(read_attribute(:preparation_steps)).map do |preparation_step|
      PreparationStep.new(preparation_step[:order_number], preparation_step[:description])
    end
  end
end